docs/documentation/design/deducer-design.zh-CN.md

```diff
@@ -0,0 +1,232 @@
+		# do something
+
+	def id() -> str: # special method
+		# do something
+		return _id
+```
+
+因此，Deducer 需要在程序代码中找到：1）特殊对象被实例化的过程（构造函数的调用过程）；2）特殊对象对特殊方法的调用过程。同时确定这些调用过程中的信息。由于，程序代码的组织形式复杂多样，这些过程可能出现在各类位置，包括嵌套函数、函数闭包、依赖库等情况，这里列举几个例子，在实现中并不一定全部支持，但需要给用户友好提示。
+
+**（1）实例化过程在闭包内**
+
+该例子中，实例化过程封装在一个函数 `createAndConfigQueue` 中，构造函数的参数由上层函数入参给定。 `createAndConfigQueue` 函数作为参数入参传入 `buildQueue` 函数中，并在其中被调用。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，两次传入的参数分别是 `queue1` 和 `queue2`，分别对应于 `queue1`、`queue2` 两个特殊对象。
+
+```typescript
+function createAndConfigQueue(name: string, kv): resource.Queue {
+  const queue = new resource.Queue(name /* other configuration */);
+  // do something
+  return queue;
+}
+
+function buildQueue(queueName: string, buildClosure: Function): resource.Queue {
+  return buildClosure(queueName);
+}
+
+const queue1 = buildQueue("queue1", createAndConfigQueue);
+const queue2 = buildQueue("queue2", createAndConfigQueue);
+```
+
+**（2）隐含特殊对象**
+
+该例子中，特殊类型的构造函数返回没有赋值给变量，而是直接调用了该返回值的特殊方法`subscribe`。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，且参数为空，这次实例化过程产生了一个特殊对象；
+2. 随后调用了这个特殊对象的 `subscribe` 特殊方法，传入此次调用的参数是一个函数；
+3. Deducer 将该函数提取为闭包函数。
+
+```typescript
+new resource.Queue().subscribe(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+**（3）函数入参做特殊方法的调用方**
+
+该例子中，特殊方法 `push` 在函数 `pushOneMessage` 被调用，而调用方，则是根据函数入参确定。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，两次传入的参数分别是 `queue1` 和 `queue2`，分别对应于 `queue1`、`queue2` 两个特殊对象。
+2. `queue.push` 特殊方法被调用一次，调用方是特殊对象 `queue1` ，传入的参数是 `"Hello, Pluto!"`。
+
+```typescript
+const queue1 = new resource.Queue("queue1");
+const queue2 = new resource.Queue("queue2");
+
+function pushOneMessage(queue: resource.Queue, msg: string) {
+  queue.push(msg);
+}
+
+pushOneMessage(queue1, "Hello, Pluto!");
+```
+
+**（4）函数参数是函数返回值**
+
+该例子中，特殊类型 `resource.Queue` 的构造函数传入的参数是 `getName()` 的返回值。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，传入的参数是 `queueName`，对应 `queue` 是一个特殊对象。
+
+```typescript
+function getName(): string {
+  return "queueName";
+}
+
+const queue = new resource.Queue(getName());
+```
+
+**（5）特殊方法重命名，间接访问**
+
+该例子中，特殊对象 `queue` 的特殊方法 `subscribe` 被赋值给 subFunc 函数变量，随用调用了该函数变量。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用一次，传入的参数是 `queue`，对应 `queue` 是一个特殊对象；
+2. 特殊方法 `subscribe` 被调用一次，调用方是特殊对象 `queue`，传入的参数是一个函数；
+3. 将该函数参数提取成闭包。
+
+```typescript
+const queue = new resource.Queue("queue");
+
+const subFunc = queue.subscribe.bind(queue);
+
+subFunc(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+**（6）动态资源对象访问，间接访问**
+
+该例子中，构建了一个 JS 对象 `queues`，该对象包含键值对，值都为特殊对象。随后，通过索引访问其中一个特殊对象。
+
+Deducer 需要确定：
+
+1. `new resource.Queue` 被调用两次，传入的参数分别是 `queue1`和 `queue2`，对应有两个特殊对象；
+2. 特殊方法 `subscribe` 被调用一次，调用方是 `queues["one"]` 对应的特殊对象，传入的参数是一个函数；
+3. 将该函数参数提取成闭包。
+
+```typescript
+const queues = {
+  one: new resource.Queue("queue1"),
+  two: new resource.Queue("queue2"),
+};
+
+queues["one"].subscribe(async () => {
+  console.log("Hello, Pluto!");
+});
+```
+
+## 所需的原子能力
```

Comment:
Are these abilities the same for different languages? Or is there a slight difference.

Reply:
The abilities for various languages should be the same, but it's uncertain whether they can be implemented in other languages.